src/python/pants/backend/python/lint/isort/subsystem.py

```diff
@@ -65,6 +65,29 @@ class Isort(PythonToolBase):
             """
         ),
     )
+    firstparty_discovery = BoolOption(
+        default=False,
+        advanced=True,
+        help=softwrap(
+            """
+            If enabled, Pants will tell `isort` explicitly which imported modules are firstparty (as
+            decided by if Pants is aware of the module via `python_` targets).
+
+            Enabling this option works around a common "gotcha" of isort runs in Pants, where isort's
+            default heuristic of "its firstparty if the file exists locally" can lead to surprising
+            behavior in Pants where the sandbox for the run is only populated with a subset of the
+            files in repo, leading to isort marking firstparty modules as thirdparty.
+
+            The downside to enabling this option is that before isort can run, Pants will be forced
+            to find the direct dependencies of every file being formatted, which can be expensive
+            in some repos (very large repos, or repos with expensive dependency-providing plugins).
```

Comment:
I don't have a good intuition for this yet, but I kinda imagine the performance might not be too bad?

- for the common(?) case (no dep-providing plugins), I _imagine_ computing deps is fast, given the new Rust parser? Way faster than running isort, anyway.
- and, generally, the dependencies likely need to be computed close to when someone runs isort too, and thus will either be cached for isort, or it's a cost that is about to be paid, e.g. running tests, building artefacts or even running mypy(?) will need to compute them?

Thoughts?

Reply:
Ah, maybe we can at least get some sort of numbers using @jriddy's https://github.com/jriddy/clam-diggers/pull/1 "benchmark" repo.

Reply:
See https://github.com/pantsbuild/pants/pull/19869#issuecomment-1728199447

Unfortunately, it had a noticeable cost in this repo :cry: We don't batch `mypy` inputs, so we'd likely memoize a little bit of the dep traversal but I wouldn't expect us to be upfront-ing any costs.